### PR TITLE
Add Union normalization parameter on spectrogram and inverse spectrogram

### PR DESCRIPTION
**Pull request by skim0514**
Add str to normalized parameter to enable frame_length based normalization to align with torch implementation of stft. Addresses issue #2104

### CODE REVIEW

**Comment by skim0514** [comment]
cc @adefossez

File: torchaudio/functional/functional.py [hunk 1]
```diff
@@ -176,10 +188,20 @@ def inverse_spectrogram(
         Tensor: Dimension `(..., time)`. Least squares estimation of the original signal.
     """
 
+    if type(normalized) is str and normalized not in ["frame_length", "window"]:
```
Comment (mthrok) [inline torchaudio/functional/functional.py:191]: To continue supporting TorchScript, need to perform type refinement.
See example like;

https://github.com/pytorch/text/blob/ca2e5a4601dcaf08bcb7171a3feebb2ecd86d230/torchtext/functional.py#L26-L38

File: torchaudio/functional/functional.py [hunk 2]
```diff
@@ -77,7 +77,8 @@ def spectrogram(
         power (float or None): Exponent for the magnitude spectrogram,
             (must be > 0) e.g., 1 for energy, 2 for power, etc.
             If None, then the complex spectrum is returned instead.
-        normalized (bool): Whether to normalize by magnitude after stft
+        normalized (bool or str): Whether to normalize by magnitude after stft. If input is str, choices are
```
Comment (mthrok) [inline torchaudio/functional/functional.py:80]: Can you update the docstrings of corresponding transforms?

Comment (nateanl) [reply torchaudio/functional/functional.py:80]: Also it's better mention if `normalized` is of `bool` type and is `True`, what's the mapped `str` option for it.

File: torchaudio/functional/functional.py [hunk 3]
```diff
@@ -162,7 +173,8 @@ def inverse_spectrogram(
         n_fft (int): Size of FFT
         hop_length (int): Length of hop between STFT windows
         win_length (int): Window size
-        normalized (bool): Whether the stft output was normalized by magnitude
+        normalized (bool or str): Whether the stft output was normalized by magnitude. If input is str, choices are
+            ``"window"`` and ``"frame_length"``, dependent on normalization mode.
```
Comment (nateanl) [inline torchaudio/functional/functional.py:177]: same here, describe the mapping between `bool` and `str`.

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 4]
```diff
@@ -245,6 +245,76 @@ def test_spectrogram_grad_at_zero(self, power):
...
+        """This test assumes that currently, torch.stft and the existing math behind spectrogram has been correct.
+        The test is checking that in relation to one another, the normalization factors correctly align based on
+        mathematical prediction. Using spec_false as a base, which has no normalization factors, we check to see that
+        turning normalized as ``True`` or ``"window"`` will have a normalization factor of sqrt(3*nfft/8) given a
+        hann_window. The sum of squares of hann has been calculated to be 3/8 of frame_length, so we are assuming this.
+        Next, when normalized ``"frame_length"`` we are using the torch stft version of normalization, therefore we
+        assume that it correctly is normalized by a factor of sqrt(nfft). This test does not test the accuracy of
+        spectrogram, but is testing the relative factors of normalization and that they align upon the frame_length
+        and chosen normalize parameter.
+        """
```
Comment (carolineechen) [inline test/torchaudio_unittest/functional/functional_impl.py:265]: can you also add the link to the github issue that brought up this problem here, provides some context on why we add this test

**Comment by carolineechen** [comment]
can you also rebase onto main for build_docs fix, or run it locally and include a screenshot?

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 5]
```diff
@@ -245,6 +245,77 @@ def test_spectrogram_grad_at_zero(self, power):
...
 
+    @parameterized.expand(
+        [
+            (1024,),
+            (2048,),
+            (4096,),
+        ]
+    )
+    def test_spectrogram_normalization_hann_window(self, nfft):
+        """This test assumes that currently, torch.stft and the existing math behind spectrogram has been correct.
```
Comment (nateanl) [inline test/torchaudio_unittest/functional/functional_impl.py:256]:
```suggestion
        """This test assumes that currently, torch.stft and the existing math behind spectrogram are correct.
```

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 6]
```diff
@@ -245,6 +245,77 @@ def test_spectrogram_grad_at_zero(self, power):
...
+        The test is checking that in relation to one another, the normalization factors correctly align based on
+        mathematical prediction. Using spec_false as a base, which has no normalization factors, we check to see that
+        turning normalized as ``True`` or ``"window"`` will have a normalization factor of sqrt(3*nfft/8) given a
+        hann_window. The sum of squares of hann has been calculated to be 3/8 of frame_length, so we are assuming this.
```
Comment (nateanl) [inline test/torchaudio_unittest/functional/functional_impl.py:260]:
```suggestion
        turning normalized as ``True`` or ``"window"`` will have a normalization factor of the sum of squares of hann
        window, which is calculated to be sqrt(3 * nfft / 8).
```

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 7]
```diff
@@ -245,6 +245,77 @@ def test_spectrogram_grad_at_zero(self, power):
...
+        Next, when normalized ``"frame_length"`` we are using the torch stft version of normalization, therefore we
```
Comment (nateanl) [inline test/torchaudio_unittest/functional/functional_impl.py:261]:
```suggestion
        Next, when ``normalized`` is ``"frame_length"``, we are using the normalization in torch.stft, therefore we
```

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 8]
```diff
@@ -245,6 +245,77 @@ def test_spectrogram_grad_at_zero(self, power):
...
+        assume that it correctly is normalized by a factor of sqrt(nfft). This test does not test the accuracy of
```
Comment (nateanl) [inline test/torchaudio_unittest/functional/functional_impl.py:262]:
```suggestion
        assume that it is correctly normalized by a factor of sqrt(nfft). This test does not test the accuracy of
```

**Review by carolineechen — APPROVED** [review]
thanks for driving this! can you address zhaoheng's suggested change as well?

**Pull request closed without merging**